CppCoreGuidelines.md

```diff
@@ -13926,29 +13926,52 @@ However, over time, code fragments can turn up in unexpected places.
 
     double cached_computation(double x)
     {
-        // bad: these two statics cause data races in multi-threaded usage
+        // bad: these statics cause data races in multi-threaded usage
         static double cached_x = 0.0;
         static double cached_result = COMPUTATION_OF_ZERO;
-        double result;
 
-        if (cached_x == x)
-            return cached_result;
-        result = computation(x);
-        cached_x = x;
-        cached_result = result;
-        return result;
+        if (cached_x != x) {
+            cached_x = x;
+            cached_result = computation(x);
+        }
+        return cached_result;
     }
 
 Although `cached_computation` works perfectly in a single-threaded environment, in a multi-threaded environment the two `static` variables result in data races and thus undefined behavior.
 
-There are several ways that this example could be made safe for a multi-threaded environment:
+##### Example, good
+
+    struct ComputationCache {
+        double cached_x = 0.0;
+        double cached_result = COMPUTATION_OF_ZERO;
+
+        double compute(double x) {
+            if (cached_x != x) {
+                cached_x = x;
+                cached_result = computation(x);
+            }
+            return cached_result;
+        }
+    };
+
+Here the cache is stored as member data of a `ComputationCache` object, rather than as shared static state.
+This refactoring essentially delegates the concern upward to the caller: a single-threaded program
+might still choose to have one global `ComputationCache`, while a multi-threaded program might
+have one `ComputationCache` instance per thread, or one per "context" for any definition of "context."
```

Comment:
I think this is a typo  
"context." --> "context".

Reply:
Not a typo. See Rule 4 here: https://www.grammarbook.com/punctuation/quotes.asp
and the first rule here: https://www.thepunctuationguide.com/quotation-marks.html

Reply:
Thank you for sharing the rule!
I am not sure actually because I am not a native English speaker
I seems I've touched on a difficult problem
Now I searched as below and found that the guide is using two things in combination, so I'm confused 

① ".
② ."

Reply:
@cugone's links are really excellent! The second one even links onward to a description specifically of `."` versus `".` style. https://www.thepunctuationguide.com/british-versus-american-style.html

In ordinary English, we never actually "quote" someone's period, comma, colon, semicolon, or dash; we quote their _words_ and sometimes their _intonation_ (question mark or exclamation point) but never their _pauses_. Commas, periods, semicolons, etc. are placed as needed on either side of the quotation without harming the sense. (American English arranges the punctuation differently in the `,"` and `."` cases for aesthetic reasons; think of it like a [ligature](https://en.wikipedia.org/wiki/Orthographic_ligature).) C++ makes it even more difficult (as a blogger, I know!) by making some ordinarily unquoted punctuation significant enough that it must sometimes be included in the quotation. 

> JavaScript programmers may be surprised to learn that `return x` is different from `return x;`.

> JavaScript programmers may be surprised to learn that `return x` is different from `return x;`; everyone is surprised to learn that `sizeof...(x)` is different from `sizeof(x)...`.

English has figured out how to deal with "I admire Martin Luther King Jr." (the periods combine), but not "`return x;`." 🙂